core/src/main/java/com/netflix/iceberg/util/BinPacking.java

```diff
@@ -107,7 +113,22 @@ public boolean hasNext() {
           bins.addLast(bin);
 
           if (bins.size() > lookback) {
-            return ImmutableList.copyOf(bins.removeFirst().items());
+            if (expensiveTaskFirst) {
+              Bin maxBin = null;
+
+              // Iterate through all bins looking for one with maximum weight, taking O(n) time.
+              for (Bin currBin : bins) {
+                maxBin = maxBin == null ? currBin : (maxBin.weight() >= currBin.weight() ? maxBin : currBin);
+              }
+              // Sanity checks: we have resolved maxBin and removed it from list of bins.
+              if (maxBin != null && bins.remove(maxBin)) {
```

Comment:
It seems to me maxBin can never be null. If that is the case then I think it is better to remove this check

Reply:
It also seems that, since maxBin is one of the elements in bin. `bins.remove(maxBin)` will always be true. So the entire if else block can be reduced to 
`ImmutableList.copyOf(bins.remove(maxBin).items())`
Which I feel is more similar to the existing code written. But I understand this specific comment is minor and could be subjective

Reply:
I agree here, too. I think this block can be simplified quite a bit.

Reply:
@rdsr - `bins.remove(maxBin)` returns a `boolean` and would not work as you've suggested unless I use the index of the bin for removal. But to your point, I'll work on making this code simpler. 

I added the `null` check in case when the head of the bin is `null` but I now realize, we might never hit this exception. I'll remove this.